Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidInvocationAsArgumentCodeFixProvider.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Text;
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Maintainability
+{
+
+	[ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(AvoidInvocationAsArgumentCodeFixProvider)), Shared]
+	public class AvoidInvocationAsArgumentCodeFixProvider : CodeFixProvider
+	{
+		private const string Title = "Extract local variable";
+
+		public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(Helper.ToDiagnosticId(DiagnosticIds.AvoidInvocationAsArgument));
+		public sealed override FixAllProvider GetFixAllProvider()
+		{
+			return WellKnownFixAllProviders.BatchFixer;
+		}
+
+		public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			Diagnostic diagnostic = context.Diagnostics.First();
+			TextSpan diagnosticSpan = diagnostic.Location.SourceSpan;
+
+			// Find the nested method call identified by the diagnostic.
+			ExpressionSyntax node = root.FindNode(diagnosticSpan, false, true) as ExpressionSyntax;
+
+			// Make sure it's part of a statement that we know how to handle.
+			// For example, we do not currently handle method expressions (ie =>)
+			StatementSyntax fullExistingExpressionSyntax = node.FirstAncestorOrSelf<ExpressionStatementSyntax>();
+			fullExistingExpressionSyntax ??= node.FirstAncestorOrSelf<ReturnStatementSyntax>();
+			fullExistingExpressionSyntax ??= node.FirstAncestorOrSelf<LocalDeclarationStatementSyntax>();
+
+			if (fullExistingExpressionSyntax != null)
+			{
+				// Register a code action that will invoke the fix.
+				context.RegisterCodeFix(
+					CodeAction.Create(
+						title: Title,
+						createChangedDocument: c => ExtractLocalVariable(context.Document, node, c),
+						equivalenceKey: Title),
+					diagnostic);
+			}
+		}
+
+		private async Task<Document> ExtractLocalVariable(Document document, ExpressionSyntax argumentSyntax, CancellationToken c)
+		{
+			SyntaxNode rootNode = await document.GetSyntaxRootAsync(c).ConfigureAwait(false);
+			const string NewName = @"renameMe";
```

Comment:
Instead of "renameMe", can't we think of something more context aware, like "resultOfFoo" (if the method is called Foo). This also avoids name clashes if the violation occurs more than once in a scope block.